### PR TITLE
Allow supressing exceptions in diamond inheritance cases

### LINKED ISSUE #16123: [Default interface methods] Constraint resolution throws ambiguous match exception too early
The exception is uncatchable because we throw it at the time of JITting the method.

We likely need a try/catch.

```
.assembly extern mscorlib { }

.assembly constrainedcall { }

.class interface private abstract auto ansi IFoo
{
  .method public hidebysig newslot abstract virtual instance void Frob() cil managed
  {
  }
}

.class interface private abstract auto ansi IBar
       implements IFoo
{
  .method public hidebysig newslot virtual final instance void Frob() cil managed
  {
    .override IFoo::Frob
    ret
  }
}

.class interface private abstract auto ansi IBaz
       implements IFoo
{
  .method public hidebysig newslot virtual final instance void Frob() cil managed
  {
    .override IFoo::Frob
    ret
  }
}

.class private auto ansi beforefieldinit Fooer
       extends [mscorlib]System.ValueType
       implements IBar, IBaz
{
}

.class private auto ansi beforefieldinit Gen`1<(IFoo) T>
       extends [mscorlib]System.Object
{
  .method public hidebysig static int32 Check(!T x) cil managed
  {
    .try
    {
      ldarga.s 0
      constrained. !T
      callvirt instance void IFoo::Frob()
      leave.s Fail
    }
    catch [mscorlib]System.NotSupportedException
    {
      pop
      leave.s Success
    }
  Fail:
    ldc.i4 -1
    ret
  Success:
    ldc.i4 100
    ret
  }
}

.method public hidebysig static int32 Main() cil managed
{
  .entrypoint
  .locals init (valuetype Fooer)
  ldloc.0
  call int32 class Gen`1<valuetype Fooer>::Check(!0)
  ret
}
```

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
This is an attempt to add a flag that allows suppressing exceptions in diamond inheritance case of default interface methods so that we don't have to wrap things in a try/catch.

Fixes #16123

### CODE REVIEW

File: src/vm/virtualcallstub.cpp [hunk 1]
```diff
@@ -2085,7 +2085,8 @@ VirtualCallStubManager::Resolver(
     MethodTable * pMT, 
     DispatchToken token,
     OBJECTREF   * protectedObj, // this one can actually be NULL, consider using pMT is you don't need the object itself
-    PCODE *       ppTarget)
+    PCODE *       ppTarget,
+    BOOL          throwOnConflict)
```
Comment (davidwrighton) [inline src/vm/virtualcallstub.cpp]: >throwOnConflict [](start = 18, length = 15)

Is this supposed to be throwOnConflict or throwOnFailure? If throwOnFailure, there is also the type equivalence case to deal with.

File: src/vm/methodtable.cpp [hunk 2]
```diff
@@ -9898,7 +9915,7 @@ MethodTable::TryResolveConstraintMethodApprox(
                 if (this->CanCastToInterface(pInterfaceMT))
                 {
                     // We can resolve to exact method
-                    pMD = this->GetMethodDescForInterfaceMethod(pInterfaceMT, pInterfaceMD);
+                    pMD = this->GetMethodDescForInterfaceMethod(pInterfaceMT, pInterfaceMD, FALSE /* throwOnConflict */);
                     _ASSERTE(pMD != NULL);
```
Comment (davidwrighton) [inline src/vm/methodtable.cpp]: Assertion no longer accurate

Comment (davidwrighton) [reply src/vm/methodtable.cpp]: fIsExactMethodResolved should be false if pMT is NULL

---
In reply to: [227113757](https://github.com/dotnet/coreclr/pull/20458#discussion_r227113757) [](ancestors = 227113757)

File: src/vm/virtualcallstub.cpp [hunk 3]
```diff
@@ -2518,13 +2527,13 @@ VirtualCallStubManager::TraceResolver(
     CONSISTENCY_CHECK(CheckPointer(pMT));
 
 
-    DispatchSlot slot(pMT->FindDispatchSlot(token));
+    DispatchSlot slot(pMT->FindDispatchSlot(token, TRUE /* throwOnConflict */));
```
Comment (davidwrighton) [inline src/vm/virtualcallstub.cpp]: throwing here is going to do something odd in the debugger when stepping into a failing interface dispatch. I'm not convinced throwing is wrong, but I'm not sure what it'll do.

File: src/vm/virtualcallstub.cpp (hunk 1 above)
Comment (davidwrighton) [reply src/vm/virtualcallstub.cpp]: Actually, I see that you handled the type equivalence case, which raises the question again on what this parameter actually means

---
In reply to: [227099736](https://github.com/dotnet/coreclr/pull/20458#discussion_r227099736) [](ancestors = 227099736)

Comment (MichalStrehovsky) [reply src/vm/virtualcallstub.cpp]: Should be `throwOnConflict`, like everywhere else. Fixed.

File: src/vm/methodtable.cpp (hunk 2 above)
Comment (MichalStrehovsky) [reply src/vm/methodtable.cpp]: Ah, yes. We would hit that assert now. Fixed.

File: src/vm/virtualcallstub.cpp (hunk 3 above)
Comment (MichalStrehovsky) [reply src/vm/virtualcallstub.cpp]: Good catch, added that to the debugger work tracking issue (#15866).

**Pull request merged**